Add CMake package config file

CJKay opened the pull request:
This change enables automatic detection and consumption of Mbed TLS library targets from within other CMake projects. By generating an `MbedTLSConfig.cmake` file, consuming projects receive a more complete view of these targets, allowing them to be used as dependencies which properly inherit the transitive dependencies of the libraries.

Resolves #298. Probably useful for #2857.

## Status

READY

## Requires Backporting

NO

## Migrations

NO

## Steps to test or reproduce

Build and install Mbed TLS as usual. Create a new CMake project, use `find_package(MbedTLS)` to create a dependency on the MBed TLS package, and set `MbedTLS_DIR` to `${MBEDTLS_INSTALL_DIR}/cmake`. You can then use the following targets from your own CMake build system:

- `MbedTLS::TLS` (TLS library)
- `MbedTLS::Crypto` (Crypto library)
- `MbedTLS::X509` (X509 library)

CJKay commented:
Third time's the charm!

ronald-cron-arm commented:
Thanks for this and the pointer to related project issues.
Few comments while doing the triage (not a review):
Without any testing in our CI we are highly likely to break this at some point. Thus please consider adding a test, probably along the lines of what is done in programs/tests/cmake_subproject.
Otherwise, we have the constraint for our CMake files to be compatible down to version 2.8.12. It is probably ok if the new feature does work with 2.8.12 but the PR should not break any support we currently have with 2.8.12, and our CI unfortunately does not test it.

CJKay commented:
> Thanks for this and the pointer to related project issues.
> Few comments while doing the triage (not a review):
> Without any testing in our CI we are highly likely to break this at some point. Thus please consider adding a test, probably along the lines of what is done in programs/tests/cmake_subproject.
> Otherwise, we have the constraint for our CMake files to be compatible down to version 2.8.12. It is probably ok if the new feature does work with 2.8.12 but the PR should not break any support we currently have with 2.8.12, and our CI unfortunately does not test it.

Hiya, I've added a couple of tests for this: one for building Mbed TLS and importing it from its build directroy, and the other for building and installing Mbed TLS and importing it from its installation directory. This *should* be compatible with CMake 2.8.12 and running the Dockerised tests locally did not seem to pull anything up, but I'm not entirely sure whether the tests I added are being run at all, so I'll await review.

gyuri-szing commented on 3rdparty/CMakeLists.txt:
```diff
@@ -1,17 +1,5 @@
-list (APPEND thirdparty_inc)
-list (APPEND thirdparty_def)
-
 execute_process(COMMAND ${MBEDTLS_PYTHON_EXECUTABLE} ${CMAKE_CURRENT_SOURCE_DIR}/../scripts/config.py -f ${CMAKE_CURRENT_SOURCE_DIR}/../include/mbedtls/config.h get MBEDTLS_ECDH_VARIANT_EVEREST_ENABLED RESULT_VARIABLE result)
```
As this is not a project, i suggest turning this file into a script. For this
1.  the file should be renamed to i.e. `3rdparty.cmake`
2. any CMakeLists.txt file to be changed to use `include(${MBEDTLS_DIR}/3rdparrty.cmake)`.

gyuri-szing replied:
This is related to #4270

CJKay replied:
Could you clarify? I don't think `CMakeLists.txt` is intended to represent only root projects - the scheme used by the [official CMake repository](https://github.com/Kitware/CMake/blob/master/Source/CMakeLists.txt) suggests as much, where subcomponents have their own `CMakeLists.txt`. In this case we want to mirror the source directory hierarchy.

gyuri-szing replied:
I answered this here: https://github.com/ARMmbed/mbedtls/issues/4270#issuecomment-810441802

CJKay replied:
I don't think it doesn't make much sense for me to introduce this scheme with this unrelated PR - particularly when it's as controversial as it is - so I'll maintain the status quo and mark these conversations as resolved so that the discussion can be continued on the dedicated issue.

gyuri-szing replied:
Ok.

gyuri-szing commented on 3rdparty/everest/CMakeLists.txt:
```diff
@@ -1,16 +1,16 @@
-list (APPEND everest_inc_public)
-list (APPEND everest_inc)
-list (APPEND everest_def)
+add_library(everest
```
Same comment for this file as for 3rdparry/CMakeLists.txt. This is not a real project and I suggest turning it to a `.cmake` file.

CJKay replied:
As above, but this would have the downside of putting compiled binaries into the incorrect build subdirectory for no real gain.

gyuri-szing replied:
" incorrect build subdirectory "
Where does the cmake documentation forbids calling add_library() from an included file? If this is a supported feature how and why is the binary directory where the file will be build incorrect?

CJKay replied:
It's not forbidden, but the most common use of `add_library` inside included scripts that I am aware of is to generate parametrised targets. In this case though there's only one `everest` library and it has a dedicated subdirectory, so add_subdirectory` seems like the perfect fit.

RE: the binary directory, `add_subdirectory` mirrors the build system hierarchy so that target locations are predictable. With `add_subdirectory`, we know that `libeverest.a` can be found in `${PROJECT_BINARY_DIR}/3rdparty/everest`, next where all of its intermediate files were generated (`${PROJECT_BINARY_DIR}/3rdparty/everest/CMakeFiles`).

Anyway, as above - discussion on this ought to happen on the dedicated thread rather than this unrelated change.

gyuri-szing commented on CMakeLists.txt:
```diff
@@ -346,3 +345,23 @@ if(ENABLE_TESTING)
+    FILE "MbedTLSTargets.cmake")
+
+# Make this package visible to the system
+export(PACKAGE MbedTLS)
```
I think it is better to avoid using the "user package registry". This is a "global" setting and the project can not really know how the integrator wishes to use it, and what side effects there can be. I don't know how this works with cross-compiled packages or even if there is any well defined cmake methodology to handle this.

CJKay replied:
[The behaviour around this](https://cmake.org/cmake/help/latest/policy/CMP0090.html#policy:CMP0090) has changed a lot over the versions, flipping between being enabled by default and disabled by default. As we're targeting CMake 2.8.12 at minimum I think the policy is "enabled by default", so what I can do is only only enable this line with CMake 3.15+, where global package exporting is disabled by default - this at least gives the user the option to export the package if they so desire with [`CMAKE_EXPORT_PACKAGE_REGISTRY`](https://cmake.org/cmake/help/latest/variable/CMAKE_EXPORT_PACKAGE_REGISTRY.html#variable:CMAKE_EXPORT_PACKAGE_REGISTRY).

gyuri-szing replied:
I see. That sounds like a good workaround.

CJKay commented:
Can't seem to reproduce this CI failure on my local machine or in Docker.

CJKay commented:
@ronald-cron-arm @gilles-peskine-arm All issues with the tests have been resolved - ready for review.

gilles-peskine-arm commented:
I know very little about cmake so I don't intend to review this.

d3zd3z reviewed (dismissed):
I'm happy with this change, and I think it is a good improvement/start at getting our Cmake usage a little more mainstream.

mstarzyk-mobica commented on library/CMakeLists.txt:
```diff
@@ -203,6 +205,10 @@ if(USE_SHARED_MBEDTLS_LIBRARY)
     target_link_libraries(${mbedcrypto_target} PUBLIC ${libs})
 
+    if(TARGET everest)
+        target_link_libraries(${mbedcrypto_static_target} PUBLIC everest)
```
did you mean `mbedcrypto_target`?

CJKay replied:
Good catch.

mstarzyk-mobica commented on programs/test/cmake_package/cmake_package.c:
```diff
@@ -0,0 +1,53 @@
+/*
+ *  Simple program to test that Mbed TLS builds correctly as a a CMake package.
```
typo `a a`

mstarzyk-mobica commented on programs/test/cmake_package/cmake_package.c:
```diff
@@ -0,0 +1,53 @@
+#else
+#include <stdio.h>
+#include <stdlib.h>
+#define mbedtls_fprintf         fprintf
```
These are not used in this sample program, why keep them in here?

CJKay replied:
These `cmake_package.c` files are copied verbatim from `cmake_subproject.c` - I won't pretend to know why these are here (I'm just the build system guy!).

gilles-peskine-arm commented:
For information, Mbed TLS 3.0 (to be released early July) will require a reasonably modern CMake version. See https://github.com/ARMmbed/mbedtls/pull/4617. Mbed TLS 2.2x will keep support for CMake 2.8.12. So it's ok to drop support for older CMake if you only want the new feature in Mbed TLS 3.0, but not if you also want to have it in the upcoming 2.2x LTS.

gilles-peskine-arm commented:
P.S. However, note that our CI currently has CMake 3.5.1, so in the near future we need to preserve compatibility with this version even if it isn't officially supported anymore.

CJKay commented:
@gilles-peskine-arm These changes are compatible with CMake 2.8.12. There's one small version check in there to prevent the package from being accidentally installed globally in pre-3.14 versions of CMake, which used to be the (highly problematic) default.

d3zd3z reviewed (approved):
LGTM

nekopsykose commented on CMakeLists.txt:
```diff
@@ -300,3 +301,37 @@ if(ENABLE_TESTING)
+install(
+    FILES "${CMAKE_CURRENT_BINARY_DIR}/cmake/MbedTLSConfig.cmake"
+          "${CMAKE_CURRENT_BINARY_DIR}/cmake/MbedTLSConfigVersion.cmake"
+    DESTINATION "cmake")
```
these destinations cause the files to be installed to `CMAKE_INSTALL_PREFIX/cmake`, which makes them put in (for example) `/usr/cmake/Mbed..`. the correct location is `${CMAKE_INSTALL_LIBDIR}/cmake/mbedtls` or similar

rzr commented:
Has port to 2.28 branch been considered ?

https://github.com/Mbed-TLS/mbedtls/issues/298#issuecomment-1694199841

gilles-peskine-arm commented:
@rzr Considered, but it would need more work. See https://github.com/Mbed-TLS/mbedtls/issues/5675#issuecomment-1111205559
